cmd/ipfs/daemon.go

```diff
@@ -258,28 +254,59 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
 	}
 
 	if gatewayMaddr != nil {
+
+		gwLis, err := manet.Listen(gatewayMaddr)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		// we might have listened to /tcp/0 - lets see what we are listing on
+		gatewayMaddr = gwLis.Multiaddr()
+
+		if writable {
+			fmt.Printf("Gateway (writable) server listening on %s\n", gatewayMaddr)
+		} else {
+			fmt.Printf("Gateway (readonly) server listening on %s\n", gatewayMaddr)
+		}
+		var opts = []corehttp.ServeOption{
+			corehttp.VersionOption(),
+			corehttp.IPNSHostnameOption(),
+			corehttp.GatewayOption(writable),
+		}
+		if rootRedirect != nil {
+			opts = append(opts, rootRedirect)
+		}
 		go func() {
-			var opts = []corehttp.ServeOption{
-				corehttp.VersionOption(),
-				corehttp.IPNSHostnameOption(),
-				corehttp.GatewayOption(writable),
-			}
-			if rootRedirect != nil {
-				opts = append(opts, rootRedirect)
-			}
-			if writable {
-				fmt.Printf("Gateway (writable) server listening on %s\n", gatewayMaddr)
-			} else {
-				fmt.Printf("Gateway (readonly) server listening on %s\n", gatewayMaddr)
-			}
-			err := corehttp.ListenAndServe(node, gatewayMaddr.String(), opts...)
+			err := corehttp.Serve(node, gwLis.NetListener(), opts...)
 			if err != nil {
 				log.Error(err)
 			}
 		}()
 	}
 
-	gateway := corehttp.NewGateway(corehttp.GatewayConfig{
+	// construct api endpoint
+	apiMaddr, err := ma.NewMultiaddr(cfg.Addresses.API)
+	if err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
+	}
+
+	apiLis, err := manet.Listen(apiMaddr)
+	if err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
+	}
+	// we might have listened to /tcp/0 - lets see what we are listing on
+	apiMaddr = apiLis.Multiaddr()
+	fmt.Printf("API server listening on %s\n", apiMaddr)
+
+	// TODO: dont persist in config - write to lock file..
+	if err := node.Repo.SetConfigKey("Addresses.API", apiMaddr.String()); err != nil {
+		res.SetError(err, cmds.ErrNormal)
+		return
+	}
```

Comment:
not super comfortable merging with this. why don't we make the change of writing the api address to a special file in the repo, _and then_ implement `/tcp/0` support